kratos/geometries/nurbs_surface_geometry.h

```diff
@@ -0,0 +1,607 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Thomas Oberbichler
+//
+//  Ported from the ANurbs library (https://github.com/oberbichler/ANurbs)
+//
+
+#if !defined(KRATOS_NURBS_SURFACE_GEOMETRY_H_INCLUDED )
+#define  KRATOS_NURBS_SURFACE_GEOMETRY_H_INCLUDED
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "geometries/geometry.h"
+
+#include "geometries/nurbs_shape_function_utilities/nurbs_surface_shape_functions.h"
+#include "geometries/nurbs_shape_function_utilities/nurbs_interval.h"
+#include "geometries/nurbs_shape_function_utilities/nurbs_utilities.h"
+
+
+
+namespace Kratos {
+
+template <int TWorkingSpaceDimension, class TPointType>
+class NurbsSurfaceGeometry : public Geometry<TPointType>
+{
+public:
+    ///@name Type Definitions
+    ///@{
+    /// Geometry as base class.
+    typedef Geometry<TPointType> BaseType;
+    typedef NurbsSurfaceGeometry<TWorkingSpaceDimension, TPointType> GeometryType;
+
+    typedef typename BaseType::IndexType IndexType;
+    typedef typename BaseType::SizeType SizeType;
+
+    typedef typename TPointType::Pointer PointPointerType;
+    typedef const PointPointerType ConstPointPointerType;
+    typedef TPointType& PointReferenceType;
+    typedef const TPointType& ConstPointReferenceType;
+
+    /** Array of counted pointers to point. This type used to hold
+        geometry's points.*/
+    typedef  typename BaseType::PointsArrayType PointsArrayType;
+    typedef  typename BaseType::CoordinatesArrayType CoordinatesArrayType;
+
+    /// Counted pointer of NurbsSurfaceGeometry
+    KRATOS_CLASS_POINTER_DEFINITION(NurbsSurfaceGeometry);
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Conctructor for B-Spline surfaces
+    NurbsSurfaceGeometry(
+        const PointsArrayType& rThisPoints,
+        const SizeType PolynomialDegreeU,
+        const SizeType PolynomialDegreeV,
+        const Vector& rKnotsU,
+        const Vector& rKnotsV)
+        : BaseType(rThisPoints, &msGeometryData)
+        , mPolynomialDegreeU(PolynomialDegreeU)
+        , mPolynomialDegreeV(PolynomialDegreeV)
+        , mKnotsU(rKnotsU)
+        , mKnotsV(rKnotsV)
+    {
+        KRATOS_DEBUG_ERROR_IF(rThisPoints.size() !=
```

Comment:
```suggestion
        KRATOS_ERROR_IF(rThisPoints.size() !=
```

this is only checked once, I would accept the performance penalty here
=> similar to the `Check` fct